### PR TITLE
Implement section_hexes in Rust with h3o

### PR DESCRIPTION
**Pull request by JeremyBYU**
## Summary

- Implement `section_hexes` method in Rust using the `h3o` crate for H3 hex cell generation
- Add direction encoding support (`:none`, `:forward`, `:both`) packed into the high bits of each cell ID
- Make elevation optional for hex building while maintaining backwards compatibility for polyline/simplify paths

NOTE: I can remove the test files if wanted!

## Backwards Compatibility

Elevation (`e` field) handling is now controlled per-caller:
- **Polyline/Simplify**: Still require elevation - points without it are skipped (unchanged behavior)
- **Hex building**: Elevation is optional - only `x` and `y` are needed

This is achieved via an internal `ElevationRequirement` enum that controls filtering behavior in `reader_to_points()`.

Read this: https://github.com/ridewithgps/ridewithgps/issues/17133 . The requirerd elevation might be a bug, but I think we should handle that later?

## Golden Tests

The `test_data/golden_hexes/` directory contains 44 test cases generated externally from the Rails Ruby implementation. These cover:
- Edge cases (empty, single point, identical points)
- All cardinal directions and complex patterns
- Gap detection at various distances
- Resolution levels 0-11
- Geographic edge cases (poles, antimeridian, null island)
- Scale testing (1k, 5k, 10k points)

## Test plan

- [x] All 44 Rust golden tests pass
- [x] Ruby integration tests verify FFI boundary
- [x] Existing polyline/simplify tests still pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: ruby_magnus_2/ext/ruby_tracklib/Cargo.toml [hunk 1]
```diff
@@ -6,10 +6,20 @@ authors = ["Dan Larkin <dan@danlarkin.org>"]
 publish = false
 
 [lib]
-crate-type = ["cdylib"]
+crate-type = ["cdylib", "rlib"]
```
Comment (JeremyBYU) [inline ruby_magnus_2/ext/ruby_tracklib/Cargo.toml:9]: It was added in the `section_hexes` commit (`4d17ec1`). The reason:

| Type   | Purpose                                                |
|--------|--------------------------------------------------------|
| cdylib | C dynamic library — what Ruby loads (`.so`)            |
| rlib   | Rust library — needed for `cargo test` and `cargo bench` |

### Why `rlib` was needed

Without it, you can’t run:

```sh
cargo test   # Golden tests
cargo bench  # Criterion benchmarks
```
These tools need to link against Rust library format, not the C dynamic library. The cdylib alone only produces the .so for Ruby, but doesn't allow Rust-side testing.

File: ruby_magnus_2/ext/ruby_tracklib/src/geometry.rs [hunk 2]
```diff
@@ -14,15 +14,7 @@ pub(crate) struct Point {
 }
 
 impl Point {
-    pub(crate) fn new(
```
Comment (JeremyBYU) [inline ruby_magnus_2/ext/ruby_tracklib/src/geometry.rs:17]: **The issue:** Benchmarks in `benches/` are treated as separate crates by Cargo. They can only access `pub` items, not `pub(crate)`.

| Visibility | Same crate | Tests (`#[cfg(test)]`) | Benchmarks (`benches/`) |
| :--- | :--- | :--- | :--- |
| `pub(crate)` | ✓ | ✓ | ✗ |
| `pub` | ✓ | ✓ | ✓ |

### Does it break anything?

*   **For Ruby:** No. Ruby only sees what magnus explicitly binds—the `pub` doesn't expose anything to Ruby.
*   **For Rust:** Technically yes—if someone depended on this crate, they'd see these internals. But since `publish = false`, nobody else uses this crate from Rust.

**Alternative:** Move benchmarks into `src/` as `#[cfg(test)]` modules, which would allow `pub(crate)`. But the current approach is simpler and harmless for an internal gem.

File: tracklib/src/read/section/encrypted.rs [hunk 3]
```diff
@@ -20,7 +20,7 @@ impl<'a> Section<'a> {
         }
     }
 
-    pub fn reader(&mut self, key_material: &[u8]) -> Result<SectionReader> {
+    pub fn reader(&mut self, key_material: &[u8]) -> Result<SectionReader<'_>> {
```
Comment (JeremyBYU) [inline tracklib/src/read/section/encrypted.rs:23]: The anonymous lifetime '_ tells the compiler: "infer this lifetime from context." In this case, it's tied to &self - the returned SectionReader lives as long as the borrow of self. 

It's like adding optional type annotations - the compiler already knew, but now humans can see it too.

**Comment by JeremyBYU** [comment]
I can drop the test files, but included them so we could at leave an idea of how it was verified.

File: ruby_magnus_2/ext/ruby_tracklib/src/geometry.rs (hunk 2 above)
Comment (kingcu) [reply ruby_magnus_2/ext/ruby_tracklib/src/geometry.rs:17]: +1 makes sense

**Review by kingcu — APPROVED** [review]
I understand the large changes now. We can drop the elevationreq aspect (later, no need now unless desired). When we were doing simplification, we considered 3d simplification, not just 2d, which I think was a big over complication. No harm in leaving the JSON test files except this large diff.

**Review by hcbarker — APPROVED** [review]
This all looks great, I think having these test files in the repo is a good idea
